### ISSUE TITLE
Tag is not losing "urgent" state

### ISSUE THREAD

**Issue by actionless** [awesomeWM/awesome#130]
After one of the patches merged yesterday tags are not losing "urgent" state after receiving it.

**Comment by actionless** [comment]
can't reproduce it no more

**Comment by blueyed** [comment]
Hmm, it's likely related to https://github.com/awesomeWM/awesome/commit/f2aa8d39edf1427a635a0be3cbc96c7b2f54eb5b.

But I cannot reproduce it, even using the default config.

> can't reproduce it no more

Ok, me neither.. :)

**Comment by actionless** [comment]
just reproduced it again

**Comment by actionless** [comment]
btw, is there any small tool which sets that urgent state on run?

like to do smth like `sleep 3 ; that-tool` and switch to another tag and wait for sleep to finish

**Comment by blueyed** [comment]
sleep 3 ; echo ^G

(where `^G` is literal (bell character), typically done through Ctrl-V Ctrl-G)

You can also use awesome-client:

    echo 'client.get()[2].urgent = true' | awesome-client

**Comment by actionless** [comment]
I will try with the first one. I don't think latter will help to reproduce
it.

**Comment by actionless** [comment]
i did not managed to reproduce it manually, but from time to time that happens in such scenarios:
1) create a rule for some app to open on tag "2"
2) go to tag "1"
3) open that app
tag "2" receives urgent state

**Comment by blueyed** [comment]
That's not a bug until there, is it?

**Comment by Elv13** [comment]
maybe the c:tags() is present before the tagged signal and that result in the counter being incremented twice? Just a theory

**Comment by actionless** [comment]
@blueyed 
> That's not a bug until there, is it?

sorry, i did not understood the question clearly

**Comment by blueyed** [comment]
> tag "2" receives urgent state

That's not a bug, is it?
Or do you mean that a tag gets marked as urgent, when a client opens there?

@psychon reported some issue on IRC where `focus = true` with `tag = ...` behaves differently now. Is this also what you mean?

Anyway, I've created a proof of concept for functional tests: https://github.com/awesomeWM/awesome/pull/133, and started adding some tests for urgent tags / clients.

**Comment by actionless** [comment]
bug is what it's not losing it. scenario was describing prerequisite, ie i am not able to reproduce the problem when tag receiving urgent state in the other way than using rules.

**Comment by blueyed** [comment]
Ok. So the bug is then that the urgent state is still there after you switch the tag?

How do you switch it? Mod4+2? Clicking the tag name?
Anything special about the client? Floating, full screen?
Is the client also marked as urgent in the tasklist, or is just the tag marked as urgent after activating the client?
What does `mod4-u` (go to next urgent) do?  Does it switch to tag 2 from somewhere else?

**Comment by blueyed** [comment]
@Elv13 
> maybe the c:tags() is present before the tagged signal and that result in the counter being incremented twice? Just a theory

Good idea/point.

@actionless 
You could add some `print` statements to the signal handlers in `lib/awful/tag.lua.in` to see what they are doing.  E.g. keep track of the `urgent_count`, and see if that's the issue.

**Comment by actionless** [comment]
>How do you switch it? Mod4+2? Clicking the tag name?
either
>Anything special about the client? Floating, full screen?
no
>Is the client also marked as urgent in the tasklist, or is just the tag marked as urgent after activating the client?
no
>What does mod4-u (go to next urgent) do? Does it switch to tag 2 from somewhere else?
i will check it
>You could add some print statements to the signal handlers in lib/awful/tag.lua.in to see what they are doing. E.g. keep track of the urgent_count, and see if that's the issue.
i will inform you when will have anything

**Comment by actionless** [comment]
just reproduced the problem again:

>What does mod4-u (go to next urgent) do? Does it switch to tag 2 from somewhere else?
if there is no really urgent client or tags then nothing happens

but changes to awful.tag (adding print statement) were rewritten when i was updating it from git, so i will add them again and will wait for problem to happen again

**Comment by blueyed** [comment]
I've merged #116, which fixes #134 (confirmed through a test - see #133).

So please try with current master and see if this might have fixed this issue, too.

**Comment by actionless** [comment]
still reproduces

**Comment by Elv13** [comment]
@actionless Can you replace:

lib/awful/ewmh.lua.in
```lua
function ewmh.urgent(c, urgent)
    if c ~= client.focus and not aclient.property.get(c,"ignore_urgent") then
        c.urgent = urgent
    end
end
```

by:

```lua
local glib = require("lgi").GLib
function ewmh.urgent(c, urgent)
    glib.idle_add(glib.PRIORITY_DEFAULT_IDLE, function()
        if c ~= client.focus and not aclient.property.get(c,"ignore_urgent") then
            c.urgent = urgent
        end
    end)
end
```

And try to reproduce?

**Comment by blueyed** [comment]
@actionless 
When it happened for you, you can also inspect the "urgent_count" for the tag using this (this is for tag 6 on screen 1):

    echo 'awful=require("awful"); t = tags[1][6]; return tostring(awful.tag.getproperty(t, "urgent_count"))'

I'd really like to have a way to reproduce this though, so it could be added as a regression test after being fixed.

I've tried it using the following rule, and with different combinations of existing clients on that tag.

    { rule = { class = "XTerm" }, properties = { tag = tags[1][6] } },

**Comment by blueyed** [comment]
@actionless 
What's the rule that you're using to move the client to another tag?

**Comment by blueyed** [comment]
It seems like the default config requires `focus=true` for the urgent flag to appear in the first place.

**Comment by Elv13** [comment]
```It seems like the default config requires focus=true for the urgent flag to appear in the first place.```
Umm, that doesn't make any sense, urgent is [supposed to be] used to warn the user when he/she is doing something else

**Comment by actionless** [comment]
https://github.com/actionless/awesome_config/blob/devel/config/rules.lua#L37

i will check `urgent_count

**Comment by Elv13** [comment]
The more I think of it, the more likely it is a race between tagged and urgent, in that case, the code in  https://github.com/awesomeWM/awesome/issues/130#issuecomment-74679371 will fix it. Maybe it would even make sense to wait 2 loop iterations.

**Comment by blueyed** [comment]
I am now writing tests for this.

What is the expected behavior for the following two cases:

1. with a rule for another tag, should the client become urgent?
2. with a rule for another tag and focus=true, should the client become urgent?

**Comment by actionless** [comment]
1) yes
2) no because urgent state should be lost on focus

**Comment by blueyed** [comment]
@actionless 
2) "on focus" is not the same as focus=true.  Because only with "switchtotag" the client would be focused always.

**Comment by actionless** [comment]
@Elv13  i applied a patch, will wait

**Comment by actionless** [comment]
@blueyed then yes, i think

**Comment by blueyed** [comment]
@actionless 
I've created a branch / PR to fix this issue: https://github.com/awesomeWM/awesome/pull/137.
For now it contains a fix for when a client gets untagged, which includes unmanaged.

@Elv13 
I don't think a delayed call is appropriate here: it should only be used when falling back to something, e.g. "tag with current" or "autofocus".
But in this case a property is requested to be set, and this needs to be handled right then.

Because the behavior of `client.focus` changed recently (it can be invisible now), the check might / should include this condition probably:

    > if c ~= client.focus and not aclient.property.get(c,"ignore_urgent") then

I will write a test for it and then see if this is required.

The current tests for the urgent property are at: https://github.com/blueyed/awesome/blob/ade5b5c1f22dae489b1740a64f87dea149c75c75/t/test-urgent.lua.
I would appreciate some review, e.g. if the assumptions are correct and how easy the test (steps) are to comprehend.

**Comment by actionless** [comment]
>i will check urgent_count when will face an issue again

it's "2"

**Comment by actionless** [comment]
and about patch in #137 -- with it i can't reproduce it at the moment, but while the problem happens from time to time mb i need more time

**Comment by blueyed** [comment]
> and about patch in #137 -- with it i can't reproduce it at the moment, but while the problem happens from time to time mb i need more time

Cool.  From `urgent_count` being `2` it really looks like it got incremented when it should have been decremented.

**Comment by actionless** [comment]
while i can't reproduce it no more i assume it was fixed in #137
